### ISSUE TITLE
Inconsistent connectivity experience when using  reverse proxy on separate host

### ISSUE THREAD

**Issue by dcarrion87** [coder/coder#4506]
For various reasons we've recently changed the architecture of the coderd instances from combined caddy reverse proxy + tailscale sidecar to separating those out on their own.

Previously:

```
[coder-workspace] --> [ts-proxy] --> tailnet --> [ts-sidecar + caddy + coderd] 
```

New set up:

```
[coder-workspace] --> [ts-proxy] --> tailnet --> [ts-sidecar + caddy] --> reverse proxy --> [coderd]
```

Since the move users can no longer connect to their workspaces from the web interface. They can still connect using `coder ssh workspace`.

If I set these environment variables on the coderd instance the web connect works but it's very slow.

```
          env {
            name  = "CODER_DERP_SERVER_ENABLE"
            value = "false"
          }
          env {
            name  = "CODER_DERP_CONFIG_URL"
            value = "https://controlplane.tailscale.com/derpmap/default"
          }
```

What would be difference about how the web interface connects users to their worksapce vs `coder ssh` to cause such an issue?

I can confirm the workspace can get to the coderd instance just fine.

I will revert the set up for now but just raising as an issue.

**Comment by dcarrion87** [comment]
Thought I would post what comes through in the logs when it "hangs" in the web terminal when using new config

>2022-10-12 10:02:27.735 [INFO]	(coderd.tailnet)	<./tailnet/conn.go:438>	(*Conn).sendNode.func1	sending node	{"node": {"id": 4010575572558331892, "key": "nodekey:XXXXXXX7ee4526ee44a6dfdc12ef10a10ca02", "disco": "discokey:XXXXXXXf70b78a29ee35d5782216c6901bd7c238f5b44b05", "preferred_derp": 0, "derp_latency": null, "addresses": ["fd7a:115c:XXX:47c1:XXX:10d8:ccf7:XXX/128"], "allowed_ips": ["fd7a:115c:XXX:47c1:XXX:10d8:ccf7:XXX/128"], "endpoints": null}}
2022-10-12 10:02:27.842 [INFO]	(coderd.tailnet)	<./tailnet/conn.go:196>	NewConn.func3	wireguard status	{"status": "\u0026{AsOf:2022-10-12 10:02:27.842407995 +0000 UTC m=+237.387316902 Peers:[] LocalAddrs:[] DERPs:1}", "err": null}
2022-10-12 10:02:27.842 [INFO]	(coderd.tailnet)	<./tailnet/conn.go:438>	(*Conn).sendNode.func1	sending node	{"node": {"id": 4010575572558331892, "key": "nodekey:XXXXXXX7ee4526ee44a6dfdc12ef10a10ca02", "disco": "discokey:XXXXXXXf70b78a29ee35d5782216c6901bd7c238f5b44b05", "preferred_derp": 0, "derp_latency": null, "addresses": ["fd7a:115c:XXX:47c1:XXX:10d8:ccf7:XXX/128"], "allowed_ips": ["fd7a:115c:XXX:47c1:XXX:10d8:ccf7:XXX/128"], "endpoints": []}}
2022-10-12 10:02:27.843 [INFO]	(coderd.tailnet)	<./tailnet/conn.go:438>	(*Conn).sendNode.func1	sending node	{"node": {"id": 4010575572558331892, "key": "nodekey:XXXXXXX7ee4526ee44a6dfdc12ef10a10ca02", "disco": "discokey:XXXXXXXf70b78a29ee35d5782216c6901bd7c238f5b44b05", "preferred_derp": 999, "derp_latency": {"999-v4": 0.002237621}, "addresses": ["fd7a:115c:XXX:47c1:XXX:10d8:ccf7:XXX/128"], "allowed_ips": ["fd7a:115c:XXX:47c1:XXX:10d8:ccf7:XXX/128"], "endpoints": []}}
2022-10-12 10:02:27.842 [INFO]	(coderd.tailnet)	<./tailnet/conn.go:196>	NewConn.func3	wireguard status	{"status": "\u0026{AsOf:2022-10-12 10:02:27.842516037 +0000 UTC m=+237.387424914 Peers:[] LocalAddrs:[{Addr:13.11.237.113:27957 Type:stun} {Addr:10.99.5.251:55724 Type:local}] DERPs:1}", "err": null}
2022-10-12 10:02:27.844 [INFO]	(coderd.tailnet)	<./tailnet/conn.go:438>	(*Conn).sendNode.func1	sending node	{"node": {"id": 4010575572558331892, "key": "nodekey:XXXXXXX7ee4526ee44a6dfdc12ef10a10ca02", "disco": "discokey:XXXXXXXf70b78a29ee35d5782216c6901bd7c238f5b44b05", "preferred_derp": 999, "derp_latency": {"999-v4": 0.002237621}, "addresses": ["fd7a:115c:XXX:47c1:XXX:10d8:ccf7:XXX/128"], "allowed_ips": ["fd7a:115c:XXX:47c1:XXX:10d8:ccf7:XXX/128"], "endpoints": ["13.11.237.113:27957", "10.99.5.251:55724"]}}

**Comment by bpmct** [comment]
@kylecarbs any ideas here?

**Comment by kylecarbs** [comment]
@dcarrion87 I believe it's because the DERP host isn't being set properly to your Tailscale DNS name, so a connection can never establish with our internal DERP.

I'm not sure how this would be though, because I _thought_ I fixed this a few weeks ago: https://github.com/coder/coder/commit/39cf329404b7f7d8277d44560be454013a27871c

Are you on v0.9.0+? If not, that's certainly worth a shot.

> What would be difference about how the web interface connects users to their worksapce vs coder ssh to cause such an issue?

It actually uses the same code, so there shouldn't be a drastic difference. Maybe it's picking a less-than-ideal DERP for some reason? But that still wouldn't make much sense. 

Let me know your version. This is certainly trivially fixable on our end once we know for certain where the issue lies!

**Comment by dcarrion87** [comment]
@kylecarbs 

- I tested this on v0.9.0 and a later version which I unfortunately did not note down.

- Your change definitely fixed my previous issues I was having with DERP. This is a new issue that presented itself after I changed the architecture to have reverse proxy + tailscale sidecar away from the coder instance itself.

I've since reverted back to the old configuration as it was eating too much time trying to investigate. It's not a huge issue for us to have coder+caddy+tailscale bunched into the same pod for now. At some point we'll want to split it out again though.
